### PR TITLE
Add acquireLoginSession API for Azdata

### PR DESCRIPTION
**Pull request by Charles-Gagnon**
This adds a new function - acquireLoginSession. When this is called it will create a session and (if there are no other active sessions) set the context such that only calls made for that specific session are allowed to execute. All other commands are queued up until the session is disposed (released) at which point it will start executing those commands (creating new sessions as requested).

There's still a lot of polish that needs to go into this - as well as tests. Getting this out now for early feedback on approach and any immediate issues seen.

This is also fixing a couple of places where we weren't passing in the additional env vars to calls such as MI edit. These additional vars need to be set to the current controller vars so that we correctly set the kube config vars when executing commands - otherwise we may end up executing commands against the wrong cluster (even if we do a login beforehand, since the login only sets the controller context. Not the kube context). 

Fixes https://github.com/microsoft/azuredatastudio/issues/13875

### CODE REVIEW

File: extensions/arc/src/models/controllerModel.ts [hunk 1]
```diff
@@ -123,64 +124,68 @@ export class ControllerModel {
...
-					}));
+		try {
+			await Promise.all([
+				this._azdataApi.azdata.arc.dc.config.show(this.azdataAdditionalEnvVars, session).then(result => {
+					this._controllerConfig = result.result;
+					this.configLastUpdated = new Date();
+					this._onConfigUpdated.fire(this._controllerConfig);
+				}).catch(err => {
+					// If an error occurs show a message so the user knows something failed but still
+					// fire the event so callers can know to update (e.g. so dashboards don't show the
```
Comment (alanrenmsft) [inline extensions/arc/src/models/controllerModel.ts:137]: I know the comment is not newly added, but I just noticed it doesn't make sense: 'can know to update'?

Comment (Charles-Gagnon) [reply extensions/arc/src/models/controllerModel.ts:137]: It's not the clearest - but it means we fire the event so that anything hooking into the event know that *something* changed and should update accordingly. 

I'll see if I can reword it to be clearer.

File: extensions/arc/src/ui/dashboards/miaa/miaaDashboardOverviewPage.ts [hunk 2]
```diff
@@ -206,8 +206,13 @@ export class MiaaDashboardOverviewPage extends DashboardPage {
 								cancellable: false
 							},
 							async (_progress, _token) => {
-								await this._controllerModel.azdataLogin();
-								return await this._azdataApi.azdata.arc.sql.mi.delete(this._miaaModel.info.name);
+								const loginSession = await this._controllerModel.acquireAzdataLoginSession();
```
Comment (alanrenmsft) [inline extensions/arc/src/ui/dashboards/miaa/miaaDashboardOverviewPage.ts:209]: the acquireAzdataLoginSession should also be placed in the try block, right? it might also fail.

File: extensions/azdata/src/typings/azdata-ext.d.ts [hunk 3]
```diff
@@ -283,13 +287,23 @@ declare module 'azdata-ext' {
...
+						loginSession?: AzdataLoginSession
 					): Promise<AzdataOutput<void>>
 				}
 			}
 		},
 		getPath(): Promise<string>,
-		login(endpoint: string, username: string, password: string, additionalEnvVars?: AdditionalEnvVars): Promise<AzdataOutput<any>>,
+		login(endpoint: string, username: string, password: string, additionalEnvVars?: AdditionalEnvVars): Promise<AzdataOutput<void>>,
+		/**
+		 * Acquires a session for the specified controller, which will log in to the specified controller and then block all other commands
```
Comment (alanrenmsft) [inline extensions/azdata/src/typings/azdata-ext.d.ts:299]: `then block all other commands from executing` is not accurate, it should be `blocking a new session from being created until...`

**Review by alanrenmsft — APPROVED** [review]

File: extensions/azdata/src/typings/azdata-ext.d.ts [hunk 4]
```diff
@@ -283,13 +287,23 @@ declare module 'azdata-ext' {
...
+		 * from executing until the session is released (disposed).
+		 * @param endpoint
+		 * @param username
+		 * @param password
+		 * @param additionalEnvVars
+		 */
+		acquireLoginSession(endpoint: string, username: string, password: string, additionalEnvVars?: AdditionalEnvVars): Promise<AzdataLoginSession>,
```
Comment (alanrenmsft) [inline extensions/azdata/src/typings/azdata-ext.d.ts:306]: I think acquireSession probably is a better name.

File: extensions/arc/src/ui/dashboards/miaa/miaaDashboardOverviewPage.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply extensions/arc/src/ui/dashboards/miaa/miaaDashboardOverviewPage.ts:209]: Yeah but if it throws then the reason for the try block (disposing the session) is unnecessary.

File: extensions/azdata/src/azdata.ts [hunk 5]
```diff
@@ -157,22 +186,68 @@ export class AzdataTool implements azdataExt.IAzdataApi {
...
+						return false;
+					}
+					return true;
+				});
+			});
+			this._queuedCommands.push({ deferred, session: undefined });
+			await deferred.promise;
+		}
+
+		await this.login(endpoint, username, password, additionalEnvVars);
```
Comment (bergeron) [inline extensions/azdata/src/azdata.ts:249]: Does this PR reduce the number of times we azdata login?  Azdata login used to be executed before each azdata call, which is unnecessary.  The logins are good for 10 hours (you can see the expiration in ~/.azdata/azdata-profile.json).  We could speed things up by either checking the expiration, or only logging in when we catch an http 401.  Preferable the later.

Comment (Charles-Gagnon) [reply extensions/azdata/src/azdata.ts:249]: No, and we can't easily reduce that since we can't know for sure that the current azdata login context is the "correct" one unless we login for every set of commands we execute (this is much more of an issue now that we support multiple controllers - although someone doing an azdata login outside of ADS would also cause this problem)

The end solution that's being worked on is to be able to set the context for a specific command through parameters passed to azdata - which would then apply to that session only. But until that work is done we need to take the perf hit and do this or we risk executing commands against the wrong cluster.
